### ISSUE TITLE
Keras - early stopping - EarlyStopping

### ISSUE THREAD

**Issue by junxnone** [junxnone/tio#377]
#9 
>EarlyStopping 是用于提前停止训练的callbacks。 当训练集上的loss不再减小（即减小的程度小于某个阈值）的时候停止训练。

**继续训练会导致测试集上的准确率下降:**
- 过拟合
- 学习率过大导致不收敛
- 使用正则项的时候，Loss的减少可能不是因为准确率增加导致的，而是因为权重大小的降低

# Reference
- [ ] [深度学习总结(十一)——early stopping](https://blog.csdn.net/manong_wxd/article/details/78745145)
- [ ] [keras中的early stopping](https://blog.csdn.net/ukakasu/article/details/80089866)
- [ ] [keras的EarlyStopping callbacks的使用与技巧](https://blog.csdn.net/silent56_th/article/details/72845912)
- [ ] [keras - callbacks - earlystopping](https://keras.io/callbacks/#earlystopping)

# API
`keras.callbacks.EarlyStopping(monitor='val_loss', min_delta=0, patience=0, verbose=0, mode='auto', baseline=None, restore_best_weights=False)`
- **monitor**: 监控的数据接口，有 `acc`, `val_acc`, `loss`, `val_loss` 等等。正常情况下如果有验证集，就用`val_acc` 或者 `val_loss`。但是因为笔者用的是5折交叉验证，没有单设验证集，所以只能用 `acc` 了。
- **min_delta**：增大或减小的阈值，只有大于这个部分才算作improvement。这个值的大小取决于monitor，也反映了你的容忍程度。例如笔者的monitor是 `acc`，同时其变化范围在70%-90%之间，所以对于小于0.01%的变化不关心。加上观察到训练过程中存在抖动的情况（即先下降后上升），所以适当增大容忍程度，最终设为0.003%。
- **patience**：能够容忍多少个epoch内都没有improvement。这个设置其实是在抖动和真正的准确率下降之间做tradeoff。如果patience设的大，那么最终得到的准确率要略低于模型可以达到的最高准确率。如果patience设的小，那么模型很可能在前期抖动，还在全图搜索的阶段就停止了，准确率一般很差。patience的大小和learning rate直接相关。在learning rate设定的情况下，前期先训练几次观察抖动的epoch number，比其稍大些设置patience。在learning rate变化的情况下，建议要略小于最大的抖动epoch number。
- **mode**: 就 `auto`, `min`, `max` 三个可能。如果知道是要上升还是下降，建议设置一下。
- **verbose**: verbosity mode.
- **baseline**: Baseline value for the monitored quantity to reach. Training will stop if the model doesn't show improvement over the baseline.
- **restore_best_weights**: whether to restore model weights from the epoch with the best value of the monitored quantity. If False, the model weights obtained at the last step of training are used.

# UseCase
```
# early stoppping
from keras.callbacks import EarlyStopping
early_stopping = EarlyStopping(monitor='val_loss', patience=50, verbose=2)
# 训练
history = model.fit(train_X, train_y, epochs=300, batch_size=20, validation_data=(test_X, test_y), verbose=2, shuffle=False, callbacks=[early_stopping])
```
